dask/dataframe/groupby.py

```diff
@@ -69,6 +69,13 @@
 #
 # #############################################
 
+SORT_SPLIT_OUT_WARNING = (
+    "In the future, `sort` for groupby operations will default to `True`"
+    " to match the behavior of pandas. However, `sort=True` does not work"
+    " with `split_out>1`. To retain the current behavior for multiple"
+    " output partitions, set `sort=False`."
+)
```

Comment:
Yeah - The `sort=True` + `split_out>1` problem is indeed the reason that we kept the `sort=None` default even after pandas moved to `sort=True`.  I think it makes sense to move away from `sort=None` support.  It also seems reasonable to adopt the `sort=True` default to be consistent with Pandas.  However, I suppose there are two significant challenges here:

1. `sort=True` is typically less performant, especially when `split_out>1`
2. `sort=True` and `split_out>1` simply don’t work together unless `shuffle=True` and there is only one groupby column

Challenge (1) certainly worries me a bit, because it seems likely that this will inevitably create a new “best practice” for Dask-Dataframe to always use  `ddf.groupby(…, `sort=False`)`.  In fact, I’m afraid we will be discussing the best way to change the default to `sort=False` in the not-too-distant future :)

Challenge (2) worries me even more, because it means that most `split_out>1` workflows will simply break without an explicit `sort=False`. This can be mitigated by using `shuffle=True` by default when `split_out>1`, but that would only cover single-column groupby aggregations (and it won’t work in `dask_cudf` until I submit a PR to add the `shuffle` code path). I suppose it should be possible to implement general-enough multi-column sorting to cover all `split_out>1` aggregations. However, the `sort=False` would certainly be faster - Taking us back to challenge (1) again.

Reply:
Thanks for taking a look @rjzamora. A couple of responses inline:


>     1. `sort=True` is typically less performant, especially when `split_out>1`
> 
>     2. `sort=True` and `split_out>1` simply don’t work together unless `shuffle=True` and there is only one groupby column
> 
> 
> Challenge (1) certainly worries me a bit, because it seems likely that this will inevitably create a new “best practice” for Dask-Dataframe to always use `ddf.groupby(…, sort=False)`. In fact, I’m afraid we will be discussing the best way to change the default to `sort=False` in the not-too-distant future :)

This PR explicitly sets `sort=False` for all intermediate partitions in the ACA-based groubpy/agg. Only the final aggregation is sorted. That is actually not a huge change from the status quo, as #9302 *accidentally* changed it from `True` to `False` by passing in a falsy value. So I'm not sure that changing the default to `True` is that huge of a deal for the `split_out = 1` case.

> Challenge (2) worries me even more, because it means that most `split_out>1` workflows will simply break without an explicit `sort=False`. This can be mitigated by using `shuffle=True` by default when `split_out>1`, but that would only cover single-column groupby aggregations (and it won’t work in `dask_cudf` until I submit a PR to add the `shuffle` code path). I suppose it should be possible to implement general-enough multi-column sorting to cover all `split_out>1` aggregations. However, the `sort=False` would certainly be faster - Taking us back to challenge (1) again.

This is entirely correct -- I was hoping that a friendly deprecation warning for a few releases would do the trick. A precedent in my mind is how the default for `pandas.DataFrame.apply` is `axis=0`. This is unsupported by Dask, but the default is still `0`, with a helpful error raised.

A more minimal version of this PR would be to just remove `None` as a valid input for `sort` and document that the default is different from the pandas default. But to me, the fact that the relative level of implicitness resulted in an unexpected change from `True` to `False` warranted some more explicitness and consistency.

Reply:
>This PR explicitly sets sort=False for all intermediate partitions in the ACA-based groubpy/agg. Only the final aggregation is sorted. That is actually not a huge change from the status quo, as #9302 accidentally changed it from True to False by passing in a falsy value. So I'm not sure that changing the default to True is that huge of a deal for the split_out = 1 case.

Yes, I totally agree that this PR is not a significant change in behavior for the `split_out=1` case. My concern is mostly for the future of the `split_out>1` case. I do think `sort=False` **should** always be a bit faster, making `sort=False` a more-favorable default for dask in the long run, but that observation doesn't mean that I think any specific change in this PR is problematic.

Reply:
> I do think sort=False should always be a bit faster

Interestingly, the performance regression from #9428 found a case where `sort=False` was *slower*, but that seems to be mostly a defect of (high-cardinality?) pandas categoricals. Mostly, I view that as a weird edge case, however.

Reply:
I wonder how valuable a more-general `sort_values` code path would be for the "doesn't work" `split_out>1` problem here? We can technically sort a DataFrame collection by an arbitrary list of columns, with arbitarary `ascending` options. The up-front calculation of the new `_partitions` column will be pretty expensive if we are sorting by multiple columns or have mixed ascending/descending columns (because we would need to actually sort each initial partition to get approximate quantiles and to assign an output partition to each row), but we **can** do it.

Reply:
> I wonder how valuable a more-general sort_values code path would be for the "doesn't work" split_out>1 problem here?

My weakly-held opinion is that sorting a groupby/agg result is not hugely important for most use-cases (this isn't really informed by any data or research, just a feeling). I'm mostly motivated by trying to be consistent and explicit here.

Reply:
>Now that we default to shuffling for split_out=1, do we expect this to be a huge problem? We're already sorting things when grouping by a single column, so the main effect here would be to show a warning/error if the user is grouping by multiple columns, advising them to set sort=False. Or do I misunderstand your concern?

I don't think you misunderstand. If we change the default to `sort=True`, then the default behavior for any user grouping by multiple columns with `split_out>1` will be to get an error advising them to add `sort=False` to the `groupby` command. I am merely pointing out that it may be frustrating for some users to get an error telling them that their default `groupby` settings are not supported. However, I'm sure all will be well as long as the error is clear and informative.

>My weakly-held opinion is that sorting a groupby/agg result is not hugely important for most use-cases (this isn't really informed by any data or research, just a feeling). I'm mostly motivated by trying to be consistent and explicit here.

I agree that most users don't actually **need** the final groups to be sorted, which is why I am a bit annoyed that the default in Pandas is the less-performant `sort=True` option :/

Reply:
> default behavior for any user grouping by multiple columns with `split_out>1` will be to get an error advising them to add `sort=False` to the `groupby` command.

We're already selecting some defaults when `split_out>1`. We could just set `True` to `False` in this case, and warn users to set it explicitly if they don't want to see a warning.

Reply:
>We're already selecting some defaults when split_out>1. We could just set True to False in this case, and warn users to set it explicitly if they don't want to see a warning.

While I like that idea, it still seems a bit confusing to enforce/document that kind of behavior with the `sort=None` option gone in the future. The underlying groupby class will need to have a `self.sort=True` or `self.sort=False` attribute, but we would be ignoring it (aside from raising a warning) for certian `split_out>1` aggregations.

Perhaps this is fine if we keep track of whether or not the `_Groupby` object was created with default `sort` attribute?

Reply:
> Perhaps this is fine if we keep track of whether or not the _Groupby object was created with default sort attribute?

How would we distinguish this in practice without keeping the default keyword argument as `None` or some `no_default` sentinel value?

I'm still not really convinced that it's onerous to ask people to set `sort=False` in some limited circumstances, i.e.
* only when `split_out > 1`
* only when grouping by multiple columns (since `shuffle` is on for `split_out > 1`)

All other cases seem to be okay to me.

Reply:
>How would we distinguish this in practice without keeping the default keyword argument as None or some no_default sentinel value?

I guess we would need to do this (or something like it) under the hood. We just wouldn't allow the user to explicitly set `sort=None`.

>I'm still not really convinced that it's onerous to ask people to set sort=False in some limited circumstances

I agree. I'm mostly taking the other side here to make sure we are doing our best to understand the implications. I think it's probably fine that users will need to set `sort=False` explicitly to enable certain `split_out>1` cases. That will break some code in the wild, but this PR provides the necessary `FutureWarning`.

It suppose the only remaining thing that "bothers" me is that the Pandas default is `sort=False`, but there is not much we can do about that :)